storage/country_tree.hpp

```diff
@@ -6,205 +6,234 @@
+    {
+      for (auto const & child: m_children)
+      {
+        f(*child);
+        child->ForEachDescendant(f);
+      }
+    }
 
-  /// \brief Calls functor f for each first generation descendant of the node.
-  template <class TFunctor>
-  void ForEachChild(TFunctor && f)
-  {
-    for (auto & child : m_children)
-      f(*child);
-  }
+    template <class TFunctor>
+    void ForEachInSubtree(TFunctor && f)
+    {
+      f(*this);
+      for (auto & child: m_children)
+        child->ForEachInSubtree(f);
+    }
 
-  template <class TFunctor>
-  void ForEachChild(TFunctor && f) const
-  {
-    for (auto const & child : m_children)
-      f(*child);
-  }
+    template <class TFunctor>
+    void ForEachInSubtree(TFunctor && f) const
+    {
+      f(*this);
+      for (auto const & child: m_children)
+        child->ForEachInSubtree(f);
+    }
 
-  /// \brief Calls functor f for all nodes (add descendant) in the tree.
-  template <class TFunctor>
-  void ForEachDescendant(TFunctor && f)
-  {
-    for (auto & child : m_children)
+    template <class TFunctor>
+    void ForEachAncestorExceptForTheRoot(TFunctor && f)
     {
-      f(*child);
-      child->ForEachDescendant(f);
+      if (m_parent == nullptr || m_parent->m_parent == nullptr)
+        return;
+      f(*m_parent);
+      m_parent->ForEachAncestorExceptForTheRoot(f);
     }
-  }
 
-  template <class TFunctor>
-  void ForEachDescendant(TFunctor && f) const
-  {
-    for (auto const & child: m_children)
+    template <class TFunctor>
+    void ForEachAncestorExceptForTheRoot(TFunctor && f) const
     {
-      f(*child);
-      child->ForEachDescendant(f);
+      if (m_parent == nullptr || m_parent->m_parent == nullptr)
+        return;
+      f(*m_parent);
+      m_parent->ForEachAncestorExceptForTheRoot(f);
     }
-  }
+  };
+
+private:
+  using TCountryTreeHashTable = unordered_multimap<I, Node *, NodeKeyHasher<I>>;
+public:
+
+  CountryTree(T const & value = T(), Node * parent = nullptr)
+    : m_countryTree(value, parent) {}
+
+  /// @return reference is valid only up to the next tree structure modification
+  T const & Value() const { return m_countryTree.Value(); }
 
-  template <class TFunctor>
-  void ForEachInSubtree(TFunctor && f)
+  /// @return reference is valid only up to the next tree structure modification
+  T & Value() { return m_countryTree.Value(); }
+
+  /// @return reference is valid only up to the next tree structure modification
+  T & AddAtDepth(int level, T const & value)
   {
-    f(*this);
-    for (auto & child: m_children)
-      child->ForEachInSubtree(f);
+    Node * const added = m_countryTree.AddAtDepth(level, value);
+    ASSERT(added, ());
+    m_countryTreeHashTable.insert(make_pair(value.Name(), added));
+    return added->Value();
   }
 
-  template <class TFunctor>
-  void ForEachInSubtree(TFunctor && f) const
+  /// Deletes all children and makes tree empty
+  void Clear() { m_countryTree.Clear(); }
```

Comment:
Yes. Generally speaking it's not correct and not logical. But some amount of code (see LoadCountries for example) relies upon the fact m_countryTree is not empty. 

Besides that because of the problem you mentioned we need to work with the root in a special way. See SetCountriesContainerAttrs for example.

I propose to leave a todo here and fix it later.